rectools/visuals/visual_app.py

```diff
@@ -122,71 +175,205 @@ def _group_interactions(
+                        request_reco,
+                        how="right",
+                        on="item_id",
+                        suffixes=["_item", "_reco"],
+                    )
+                prepared_model_reco[request_name] = request_reco
+            prepared_reco[model_name] = prepared_model_reco
+        return prepared_reco
+
+    @classmethod
+    def _ungroup_reco(
+        cls,
+        grouped_reco: tp.Dict[tp.Hashable, TablesDict],
+        selected_requests: tp.Dict[tp.Hashable, ExternalId],
+        id_col: str,
+    ) -> pd.DataFrame:
+        res = []
+        for model_name, prepared_model_reco in grouped_reco.items():
+            for request_name, request_reco in prepared_model_reco.items():
+                df = request_reco.copy()
+                df[id_col] = selected_requests[request_name]
+                df[Columns.Model] = model_name
+                res.append(df)
+        return pd.concat(res, axis=0, sort=False).reset_index(drop=True)
+
+    @classmethod
+    def _ungroup_interactions(
+        cls,
+        grouped_interactions: TablesDict,
+        selected_requests: tp.Dict[tp.Hashable, ExternalId],
+        id_col: str,
+    ) -> pd.DataFrame:
+        res = []
+        for request_name, request_interactions in grouped_interactions.items():
+            df = request_interactions.copy()
+            df[id_col] = selected_requests[request_name]
+            res.append(df)
+        return pd.concat(res, axis=0, sort=False).reset_index(drop=True)
+
+    @classmethod
+    def _check_columns_present_in_reco(cls, reco: TablesDict, id_col: str) -> None:
+        required_columns = {id_col, Columns.Item}
+        for model_name, model_reco in reco.items():
+            actual_columns = set(model_reco.columns)
             if not actual_columns >= required_columns:
                 raise KeyError(f"Missed columns {required_columns - actual_columns} in {model_name} recommendations df")
 
-    def _prepare_interactions_for_i2i(self) -> pd.DataFrame:
+    @classmethod
+    def _prepare_interactions_for_i2i(cls, reco: TablesDict) -> pd.DataFrame:
         request_ids = set()
-        for recos_df in self.recos.values():
-            request_ids.update(set(recos_df[Columns.TargetItem].unique()))
+        for reco_df in reco.values():
+            request_ids.update(set(reco_df[Columns.TargetItem].unique()))
         interactions = pd.DataFrame({Columns.TargetItem: list(request_ids), Columns.Item: list(request_ids)})
         return interactions
 
+    @classmethod
+    def _df_to_tables_dict(cls, df: pd.DataFrame, key_col: str) -> TablesDict:
+        res = {}
+        for key, grouped_df in df.groupby(key_col):
+            res[key] = grouped_df.drop(columns=[key_col]).reset_index(drop=True)
+        return res
+
+    @classmethod
+    def _create_requests_df(cls, selected_requests: tp.Dict[tp.Hashable, ExternalId]) -> pd.DataFrame:
+        df = pd.DataFrame(
```

Comment:
You can do it like this, but I recommend to use smth like
```
pd.Series(selected_requests).to_frame().rename(columns={...})
```

Last Python versions keep dict order, but still I think it's not the best practice to rely on it

Reply:
fixed